platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,754 @@
+                            needs_wake = true;
+                            State::Idle
+                        }
+                        SmhcOp::Read { .. } | SmhcOp::Write { .. } => State::WaitForDma,
+                    }
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForDma => {
+                // TODO: better way to check for RX and TX,
+                // *normal interrupt summary* does not seem to work
+                if smhc.smhc_idst.read().bits() != 0 {
+                    smhc.smhc_idst.write(|w| {
+                        w.nor_int_sum().set_bit();
+                        w.rx_int().set_bit();
+                        w.tx_int().set_bit();
+                        w
+                    });
+                    State::WaitForDataTransfer
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForDataTransfer => {
+                if smhc.smhc_mintsts.read().m_dtc_int().bit_is_set() {
+                    smhc.smhc_rintsts.write(|w| w.dtc().set_bit());
+                    match &mut self.op {
+                        SmhcOp::None | SmhcOp::Control => State::Idle,
+                        SmhcOp::Read {
+                            auto_stop,
+                            buf,
+                            cnt,
+                        }
+                        | SmhcOp::Write {
+                            auto_stop,
+                            buf,
+                            cnt,
+                        } => {
+                            tracing::trace!("setting buf len: {cnt}");
+                            // TODO: Safety?
+                            unsafe { buf.as_vec_mut().set_len(*cnt) };
+                            if *auto_stop {
+                                State::WaitForAutoStop
+                            } else {
+                                needs_wake = true;
+                                State::Idle
+                            }
+                        }
+                    }
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForAutoStop => {
+                if smhc.smhc_mintsts.read().m_acd_int().bit_is_set() {
+                    smhc.smhc_rintsts.write(|w| w.acd().set_bit());
+                    needs_wake = true;
+                    State::Idle
+                } else {
+                    self.state
+                }
+            }
+        };
+
+        if needs_wake {
+            if let Some(waker) = self.waker.take() {
+                waker.wake();
+                // If we are waking the driver task, we need to disable interrupts
+                smhc.smhc_ctrl.modify(|_, w| w.ine_enb().disable());
+            }
+        }
+    }
+}
+
+/// Internal DMA controller
+mod idmac {
+    #[repr(C, align(4))]
+    #[derive(Default, Debug)]
+    pub struct Descriptor {
```

Comment:
nit, TIOLI: this struct (and everything else in the `idmac` module) really only needs to be 
```suggestion
    pub(super) struct Descriptor {
```
since it's never re-exported...